src/flp/types/l1boundsum.rs

```diff
@@ -0,0 +1,402 @@
+
+    /// Computes the largest field element for the chosen bit count.
+    fn largest_field_element(&self) -> F::Integer {
+        let one = F::Integer::from(F::one());
+        (one << self.bits) - one
+    }
+}
+
+impl<F: NttFriendlyFieldElement, S> Clone for L1BoundSum<F, S> {
+    fn clone(&self) -> Self {
+        Self {
+            measurement_len: self.measurement_len,
+            bits: self.bits,
+            measurement_len_in_bits: self.measurement_len_in_bits,
+            chunk_length: self.chunk_length,
+            gadget_calls: self.gadget_calls,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<F, S> Flp for L1BoundSum<F, S>
+where
+    F: NttFriendlyFieldElement,
+    S: ParallelSumGadget<F, Mul<F>> + Eq + 'static,
+{
+    type Field = F;
+
+    fn gadget(&self) -> Vec<Box<dyn Gadget<F>>> {
+        vec![Box::new(S::new(
+            Mul::new(self.gadget_calls),
+            self.chunk_length,
+        ))]
+    }
+
+    fn num_gadgets(&self) -> usize {
+        1
+    }
+
+    fn valid(
+        &self,
+        gadget: &mut Vec<Box<dyn Gadget<F>>>,
+        input: &[F],
+        joint_randomness: &[F],
+        num_shares: usize,
+    ) -> Result<Vec<F>, FlpError> {
+        self.valid_call_check(input, joint_randomness)?;
+
+        let range_check = parallel_sum_range_checks(
+            &mut gadget[0],
+            input,
+            joint_randomness,
+            self.chunk_length,
+            num_shares,
+        )?;
+
+        // The encoded input consists of the measurement with a single field element appended,
+        // representing the claimed weight. The observed weight is obtained by summing up all the
+        // input elements but the last one.
+        let mut observed_weight = F::zero();
+        let mut claimed_weight = F::zero();
+        for (index, element) in input.chunks(self.bits).enumerate() {
+            let decoded = F::decode_bitvector(element)?;
+            if index == self.measurement_len {
+                // Last element: this is the claimed weight.
+                claimed_weight += decoded;
+            } else {
+                observed_weight += decoded;
+            }
+        }
+
+        Ok(vec![range_check, observed_weight - claimed_weight])
+    }
+
+    fn input_len(&self) -> usize {
+        self.measurement_len_in_bits
+    }
+
+    fn proof_len(&self) -> usize {
+        (self.chunk_length * 2) + 2 * ((1 + self.gadget_calls).next_power_of_two() - 1) + 1
+    }
+
+    fn verifier_len(&self) -> usize {
+        2 + self.chunk_length * 2
+    }
+
+    fn joint_rand_len(&self) -> usize {
+        self.gadget_calls
+    }
+
+    fn eval_output_len(&self) -> usize {
+        2
+    }
+
+    fn prove_rand_len(&self) -> usize {
+        self.chunk_length * 2
+    }
+}
+
+impl<F, S> Type for L1BoundSum<F, S>
+where
+    F: NttFriendlyFieldElement,
+    S: ParallelSumGadget<F, Mul<F>> + Eq + 'static,
+{
+    /// The measurement is the input with the L1 norm appended.
```

Comment:
This is incorrect, because the `Measurement` type comes before the AFE encoding. This is just the original vector of integers.